### PR TITLE
Move public definitions to `opentelemetry_api`.

### LINKED ISSUE #1311: We should move all public definitions in `CMakeLists.txt` to `target_compile_definitions(opentelemetry_api INTERFACE ...)`
**Describe your environment**

System: Windows
Build System: cmake

**Steps to reproduce**

`cmake <...> -DWITH_ETW=ON`

**What is the expected behavior?**

Now, some definitions are added by `add_definitions(...)` in `CMakeLists.txt` and so that these definitions are private. But we should keep the same behaviour between compiling opentelemetry-cpp and using it by users. We can add these public definitions by `target_compile_definitions(opentelemetry_api INTERFACE ...)` and then these definitions will be auto passed to all cmake targets depend `opentelemetry_api`.

**What is the actual behavior?**

Some definitions are missing and library get a different behavior.

**Additional context**

Should we add prefix to `NO_GETENV` and `NOMINMAX` in case of them will conflict with other libraries?

It this's acceptable, it can be assigned to me please.

### PR DESCRIPTION
**Pull request by owent**
Signed-off-by: owent <admin@owent.net>

Fixes #1311 

## Changes

Move definitions to `opentelemetry_api` and set them as `INTERFACE` definitions.

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [ ] Unit tests have been added
* [x] Changes in public API reviewed

### CODE REVIEW

**Review by lalitb — APPROVED** [review]
LGTM, Thanks for the cleanup.

File: CMakeLists.txt [hunk 1]
```diff
@@ -125,24 +125,11 @@ if(NOT DEFINED CMAKE_CXX_STANDARD)
   endif()
 endif()
 
-if(WITH_ABSEIL)
-  find_package(absl CONFIG REQUIRED)
-
-  set(CORE_RUNTIME_LIBS absl::bad_variant_access absl::any absl::base
```
Comment (ThomsonTan) [inline CMakeLists.txt:131]: Where is the definition of `CORE_RUNTIME_LIBS` moved to?

File: CMakeLists.txt [hunk 2]
```diff
@@ -215,17 +180,8 @@ option(
 option(WITH_EXAMPLES "Whether to build examples" ON)
 
 option(WITH_METRICS_PREVIEW "Whether to build metrics preview" OFF)
-
-if(WITH_METRICS_PREVIEW)
-  add_definitions(-DENABLE_METRICS_PREVIEW)
```
Comment (ThomsonTan) [inline CMakeLists.txt:220]: Should this and `ENABLE_LOGS_PREVIEW` below be moved to api/CMakeLists.txt as well?

File: CMakeLists.txt (hunk 1 above)
Comment (lalitb) [reply CMakeLists.txt:131]: The variable is actually removed, and these libraries are now directly linked to `opentelemetry-api`

test

File: CMakeLists.txt (hunk 2 above)
Comment (owent) [reply CMakeLists.txt:220]: These two definitions are already in `api/CMakeLists.txt`. It's duplicate here.

**Review by ThomsonTan — APPROVED** [review]

**Comment by lalitb** [comment]
@owent - Can you rebase it enable merge. Thanks

**Comment by owent** [comment]
> Reference

Done

**Pull request merged**